internal/cache/file/downloader/parallel_downloader.go

```diff
@@ -0,0 +1,196 @@
+		return
+	}
+	defer func() {
+		if err != nil {
+			_ = newReader.Close()
+			_ = cacheFile.Close()
+			return
+		}
+		err = newReader.Close()
+		if err != nil {
+			logger.Errorf("Job:%p (%s:/%s) error while closing reader: %v", job, job.bucket.Name(), job.object.Name, err)
+		}
+		err = cacheFile.Close()
+		if err != nil {
+			return
+		}
+	}()
+
+	_, err = cacheFile.Seek(start, 0)
+	if err != nil {
+		err = fmt.Errorf("downloadObjectAsync: error while seeking file handle, seek %d: %w", start, err)
+		return
+	}
+	// Copy the contents from NewReader to cache file.
+	_, readErr := io.CopyN(cacheFile, newReader, end-start)
+	if readErr != nil {
+		err = fmt.Errorf("downloadObjectAsync: error at the time of copying content to cache file %w", readErr)
+		return
+	}
+	return
+}
+
+// downloadObjectAsync downloads the backing GCS object into a file as part of
+// file cache using NewReader method of gcs.Bucket.
+//
+// Note: There can only be one async download running for a job at a time.
+// Acquires and releases LOCK(job.mu)
+func (job *Job) downloadObjectInParallelAsync() {
+	// Close the job.doneCh, clear the cancelFunc & cancelCtx and call the
+	// remove job callback function in any case - completion/failure.
+	defer func() {
+		job.cancelFunc()
+		close(job.doneCh)
+
+		job.mu.Lock()
+		if job.removeJobCallback != nil {
+			job.removeJobCallback()
+			job.removeJobCallback = nil
+		}
+		job.cancelCtx, job.cancelFunc = nil, nil
+		job.mu.Unlock()
+	}()
+
+	// Create, open and truncate cache file for writing object into it.
+	cacheFile, err := cacheutil.CreateFile(job.fileSpec, os.O_TRUNC|os.O_WRONLY)
+	if err != nil {
+		err = fmt.Errorf("downloadObjectAsync: error in creating cache file: %w", err)
+		job.failWhileDownloading(err)
+		return
+	}
+	// Assumption: If we don't have the file already created then it may happen
+	// that parallel go routines writing to the same file even at disjoint parts
+	// can make the data corrupt.
+	err = cacheFile.Truncate(int64(job.object.Size))
+	if err != nil {
+		err = fmt.Errorf("downloadObjectAsync: error while truncating the cache file: %w", err)
+		job.failWhileDownloading(err)
+		return
+	}
+	err = cacheFile.Close()
+	if err != nil {
+		err = fmt.Errorf("downloadObjectAsync: error while closing cache file: %w", err)
+		job.failWhileDownloading(err)
+	}
+
+	notifyInvalid := func() {
+		job.mu.Lock()
+		job.status.Name = Invalid
+		job.notifySubscribers()
+		job.mu.Unlock()
+	}
+
+	var start, end int64
+	end = int64(job.object.Size)
+	var readRequestSize int64 = int64(job.readRequestSizeMb * cacheutil.MiB)
+	var oneStepTotalReadSize int64 = readRequestSize * int64(job.downloadParallelism)
+
+	for {
+		select {
+		case <-job.cancelCtx.Done():
+			return
+		default:
+			if start < end {
+				// Parallely download different ranges not more than ReadChunkSize and
+				// not using go routines more than AsyncDownloadParallelism
+				asyncDownloaderErrors := make([]error, job.downloadParallelism)
```

Comment:
Instead, we should use ErrGroup: https://medium.com/@yardenlaif/go-sync-or-go-home-errgroup-f91a0ee72d3f
(a) No need to maintain error of all the go-routines.
(b) Abort if any of the go routine returns error and returns the first error (if multiple go routines returns error).